src/main/java/graphql/analysis/QueryTraverser.java

```diff
@@ -51,32 +51,38 @@ public class QueryTraverser {
                            String operation,
-                           Map<String, Object> variables) {
-        assertNotNull(document, () -> "document can't be null");
+                           CoercedVariables coercedVariables) {
+        this.schema = schema;
         NodeUtil.GetOperationResult getOperationResult = NodeUtil.getOperation(document, operation);
-        List<VariableDefinition> variableDefinitions = getOperationResult.operationDefinition.getVariableDefinitions();
-        this.schema = assertNotNull(schema, () -> "schema can't be null");
         this.fragmentsByName = getOperationResult.fragmentsByName;
         this.roots = singletonList(getOperationResult.operationDefinition);
         this.rootParentType = getRootTypeFromOperation(getOperationResult.operationDefinition);
-        this.coercedVariables = coerceVariables(assertNotNull(variables, () -> "variables can't be null"), variableDefinitions);
+        this.coercedVariables = coercedVariables;
     }
 
-    private CoercedVariables coerceVariables(Map<String, Object> rawVariables, List<VariableDefinition> variableDefinitions) {
-        return new ValuesResolver().coerceVariableValues(schema, variableDefinitions, new RawVariables(rawVariables));
+    private QueryTraverser(GraphQLSchema schema,
+                           Document document,
+                           String operation,
+                           RawVariables rawVariables) {
+        this.schema = schema;
+        NodeUtil.GetOperationResult getOperationResult = NodeUtil.getOperation(document, operation);
+        List<VariableDefinition> variableDefinitions = getOperationResult.operationDefinition.getVariableDefinitions();
+        this.fragmentsByName = getOperationResult.fragmentsByName;
+        this.roots = singletonList(getOperationResult.operationDefinition);
+        this.rootParentType = getRootTypeFromOperation(getOperationResult.operationDefinition);
+        this.coercedVariables = new ValuesResolver().coerceVariableValues(schema, variableDefinitions, rawVariables);
     }
 
     private QueryTraverser(GraphQLSchema schema,
                            Node root,
                            GraphQLCompositeType rootParentType,
                            Map<String, FragmentDefinition> fragmentsByName,
-                           Map<String, Object> variables) {
-        this.schema = assertNotNull(schema, () -> "schema can't be null");
-        assertNotNull(root, () -> "root can't be null");
+                           CoercedVariables coercedVariables) {
+        this.schema = schema;
         this.roots = Collections.singleton(root);
-        this.rootParentType = assertNotNull(rootParentType, () -> "rootParentType can't be null");
-        this.fragmentsByName = assertNotNull(fragmentsByName, () -> "fragmentsByName can't be null");
-        this.coercedVariables = new CoercedVariables(assertNotNull(variables, () -> "variables can't be null"));
```

Comment:
are the asserts removed because these invariants are no longer true?

Reply:
I'm adding 2 new constructors so to keep the code tidier I thought I would move all the null checks into the builder, rather than having to remember 4 lots of asserts